File: dotnet/CardsProject/Domain/Marketplace/TradeTransaction.cs
using CardsProject.Domain.Players;
using System.ComponentModel.DataAnnotations.Schema;
using System.ComponentModel.DataAnnotations;
using System.Text.Json.Serialization;

namespace CardsProject.Domain.Marketplace;

public enum TradeTransactionStatusType
{
    Pending,
    Completed,
    Disputed,
    Refunded
}

public class TradeTransaction : IValidatableObject
{
    public int Id { get; set; }

    public decimal FinalPrice { get; set; } = 0.00m;
    public decimal PlatformFee { get; set; } = 0.00m;
    public TradeTransactionStatusType Status { get; set; }
    [JsonPropertyName("completedAt")]
    public DateTime? CompletedAt { get; set; } = null;

    public int? ListingId { get; set; }
    [ForeignKey(nameof(ListingId))]
    public TradeListing? Listing { get; set; }
    public int? BuyerId { get; set; }
    [ForeignKey(nameof(BuyerId))]
    public Player? Buyer { get; set; }
    public int? SellerId { get; set; }
    [ForeignKey(nameof(SellerId))]
    public Player? Seller { get; set; }

    // Business operations

    public void Complete()
    {
        // TODO: implement complete
    }

    public void Refund()
    {
        // TODO: implement refund
    }

    public void OpenDispute(string reason)
    {
        // TODO: implement open_dispute
    }

    public decimal SellerNet()
    {
        // TODO: implement seller_net
        return default;
    }

    // ── Domain invariants (simple rules) ──────────────────────────────
    public IEnumerable<ValidationResult> Validate(ValidationContext validationContext)
    {
        if (!( PlatformFee <= FinalPrice ))
            yield return new ValidationResult("Platform fee cannot exceed the final price", new[] { nameof(Id) });
        if (!( PlatformFee >= 0m ))
            yield return new ValidationResult("Platform fee must not be negative", new[] { nameof(Id) });
        if (!( FinalPrice > 0m ))
            yield return new ValidationResult("Transaction final price must be greater than zero", new[] { nameof(Id) });
    }
}

File: laravel/app/Models/Marketplace/TradeTransaction.php
<?php

namespace App\Models\Marketplace;

use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\HasMany;
use Illuminate\Database\Eloquent\Relations\BelongsToMany;
use App\Models\Players\Player;

class TradeTransaction extends Model
{
    protected $table = 'trade_transactions';

    protected $fillable = ['final_price', 'platform_fee', 'status', 'completed_at', 'listing_id', 'buyer_id', 'seller_id'];

    protected $casts = [
        'final_price' => 'decimal:2',
        'platform_fee' => 'decimal:2',
        'completed_at' => 'datetime',
    ];

    const STATUS_VALUES = ['Pending', 'Completed', 'Disputed', 'Refunded'];

    public function listing(): BelongsTo
    {
        return $this->belongsTo(TradeListing::class, 'listing_id');
    }

    public function buyer(): BelongsTo
    {
        return $this->belongsTo(Player::class, 'buyer_id');
    }

    public function seller(): BelongsTo
    {
        return $this->belongsTo(Player::class, 'seller_id');
    }

    // ── Validation rules ─────────────────────────────────────────────
    public function validateRules(): void
    {
        $errors = [];
        if (!(($this->platform_fee === null || ($this->final_price !== null && (float)$this->platform_fee <= (float)$this->final_price)))) {
            $errors['fee_not_exceed_price'] = 'Platform fee cannot exceed the final price';
        }
        if (!(($this->platform_fee === null || (float)$this->platform_fee >= (float)0))) {
            $errors['fee_not_negative'] = 'Platform fee must not be negative';
        }
        if (!(($this->final_price === null || (float)$this->final_price > (float)0))) {
            $errors['final_price_positive'] = 'Transaction final price must be greater than zero';
        }
        if (!empty($errors)) {
            throw new \Illuminate\Validation\ValidationException(
                \Illuminate\Support\Facades\Validator::make([], []),
                response()->json(['errors' => $errors], 422)
            );
        }
    }

    // ── Domain invariants (IMPLIES rules) ───────────────────────────────
    public function validateImplies(): void
    {
        if ($this->status === 'Completed' && $this->completed_at === null) {
            throw new \RuntimeException('Completed transaction must have a completed_at timestamp');
        }
    }

    // ── Business operations ──────────────────────────────────────────

    public function complete(): void
    {
        throw new \RuntimeException('complete not implemented');
    }

    public function refund(): void
    {
        throw new \RuntimeException('refund not implemented');
    }

    public function openDispute($reason): void
    {
        throw new \RuntimeException('open_dispute not implemented');
    }

    public function sellerNet(): string
    {
        throw new \RuntimeException('seller_net not implemented');
    }

}
